hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/utils/BulkInsertFunctionWrapper.java

```diff
@@ -0,0 +1,232 @@
+    boolean needFixedFileIdSuffix = OptionsResolver.isNonBlockingConcurrencyControl(conf);
+    this.bucketIdToFileId = new HashMap<>();
+    this.mapFunction = r -> BucketBulkInsertWriterHelper.rowWithFileId(bucketIdToFileId, keyGen, r, indexKeys, numBuckets, needFixedFileIdSuffix);
+    if (asyncClustering) {
+      clusteringFunctionWrapper.openFunction();
+    }
+  }
+
+  public void invoke(I record) throws Exception {
+    RowData recordWithFileId = mapFunction.map((RowData) record);
+    writeFunction.processElement(recordWithFileId, null, null);
+  }
+
+  public WriteMetadataEvent[] getEventBuffer() {
+    return this.coordinator.getEventBuffer();
+  }
+
+  public OperatorEvent getNextEvent() {
+    return this.gateway.getNextEvent();
+  }
+
+  public void checkpointFunction(long checkpointId) {
+    // Do nothing
+  }
+
+  @Override
+  public void endInput() {
+    writeFunction.endInput();
+    if (bucketIdToFileId != null) {
+      this.bucketIdToFileId.clear();
+    }
+  }
+
+  public void checkpointComplete(long checkpointId) {
+    stateInitializationContext.getOperatorStateStore().checkpointSuccess(checkpointId);
+    coordinator.notifyCheckpointComplete(checkpointId);
+    if (asyncClustering) {
+      try {
+        clusteringFunctionWrapper.cluster(checkpointId);
+      } catch (Exception e) {
+        throw new HoodieException(e);
+      }
+    }
+  }
+
+  public void coordinatorFails() throws Exception {
+    this.coordinator.close();
+    this.coordinator.start();
+    this.coordinator.setExecutor(new MockCoordinatorExecutor(coordinatorContext));
+  }
+
+  public void checkpointFails(long checkpointId) {
+    coordinator.notifyCheckpointAborted(checkpointId);
+  }
+
+  public StreamWriteOperatorCoordinator getCoordinator() {
+    return coordinator;
+  }
+
+  public MockOperatorCoordinatorContext getCoordinatorContext() {
+    return coordinatorContext;
+  }
+
+  @Override
+  public void close() throws Exception {
+    this.coordinator.close();
+    if (clusteringFunctionWrapper != null) {
+      clusteringFunctionWrapper.close();
+    }
+    if (bucketIdToFileId != null) {
+      this.bucketIdToFileId.clear();
+    }
+  }
+
+  // -------------------------------------------------------------------------
+  //  Utilities
+  // -------------------------------------------------------------------------
+
+  private void setupWriteFunction() {
+    ExecutorService executors = Executors.newFixedThreadPool(2);
+    Runnable writeFunctionInitializer = () -> {
+      try {
+        writeFunction = new BulkInsertWriteFunction<>(conf, rowType);
+        writeFunction.setRuntimeContext(runtimeContext);
+        writeFunction.setOperatorEventGateway(gateway);
+        writeFunction.open(conf);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    };
+    executors.submit(writeFunctionInitializer);
+    Runnable coordinatorInitializer = () -> {
+      while (true) {
```

Comment:
So we have thread safety issue here?

Reply:
No.
I add here to avoid blocking in tests. And It is only necessary for test cases.
 `BulkInsertWriteFunction#open` would blocking until `StreamWriteOperatorCoordinator`  initialize the instant. And `StreamWriteOperatorCoordinator` would waiting for the bootstrap message from  `BulkInsertWriteFunction`. 
<img width="1225" alt="image" src="https://github.com/apache/hudi/assets/1525333/7ef0bff7-f330-4f87-bdb5-ff285d13859e">
<img width="1240" alt="image" src="https://github.com/apache/hudi/assets/1525333/76635c3a-edeb-4630-8fdf-9168d07fb44f">

Reply:
We should never block in the `#open` method.

Reply:
Add a jira to track this [HUDI-6968](https://issues.apache.org/jira/browse/HUDI-6968).
Since it's not related to this PR. And could be solved in next PR.